### PR TITLE
ENH: F2PY refactor

### LINKED ISSUE #20480: ENH: Refactor the F2PY module
### Proposed new feature or change:

The current file structure of F2PY is flat, which means that the different parts of the module are not easily distinguishable.

The key segments of the module being refactored into:
- A frontend (parser)
  - The `crackfortran` tokenizer
  - The CLI (currently `f2py2e`)
- Transforms of the AST (middle end)
  - This is the standards module --> `f77`, `f90`, `pyf` and other functions 
- Production rules (back end)
  - Code generation
- C wrapper helpers 
- Utilities

This structure would allow new contributors to say, add an `argparse` frontend without going over the rest of code. Currently only end-to-end contributions are required but the codebase is largely modular (e.g. the new symbolic solver).

A new logical structure around the lines of the functionality described would be best.

### LINKED ISSUE #20560: ENH,MAINT: Add a shim for F2PY refactors
Instead of breaking all existing F2PY dependencies with #20481, consider adding to the `__init__.py` so downstream projects are not bothered.

### PR DESCRIPTION
**Pull request by HaoZeke**
This essentially is the bare minimum set of changes required to use a the new logical structure.
- Use a new package layout
- Minimal `distutils` changes
- Reduced cruft in `__init__.py`
- Switched to absolute imports everywhere

```
f2py/frontend/crackfortran.py
f2py/frontend/f2py2e.py
f2py/codegen/rules.py
f2py/codegen/cfuncs.py
f2py/codegen/func2subr.py
f2py/utils/diagnose.py
f2py/utils/npdist.py
f2py/utils/pathhelper.py
f2py/stds/f77/common_rules.py
f2py/stds/f90/f90mod_rules.py
f2py/stds/f90/use_rules.py
f2py/stds/pyf/capi_maps.py
f2py/stds/pyf/cb_rules.py
f2py/stds/auxfuncs.py
f2py/stds/symbolic.py
f2py/csrcs/fortranobject.c
f2py/csrcs/fortranobject.h
```

# List of changes
- [x] Refactor
- [x] Add a module helper
- [x] Ensure compatibility
- [x] Test module helper 

# Rationale
- Closes #20480
- Closes #20560

### CODE REVIEW

**Comment by HaoZeke** [comment]
Note that this PR would break every existing `f2py` PR, but if it looks good I can go over and rebase all open PRs.

**Comment by HaoZeke** [comment]
Local reproducer:
```
python runtests.py -t numpy/tests/test_public_api.py
python runtests.py  -t numpy/typing/tests/test_typing.py::test_code_runs[modules] -m full
```

**Comment by HaoZeke** [comment]
AFAIK the current build error seems to be a bit of a fluke (CI timeout)

**Comment by melissawm** [comment]
There's a few conflicts now, can you check @HaoZeke ?

**Comment by HaoZeke** [comment]
> There's a few conflicts now, can you check @HaoZeke ?

Should be good to go ^_^

**Comment by charris** [comment]
@HaoZeke Needs another rebase.

**Comment by HaoZeke** [comment]
I'm not sure this should be merged as is and followed up with:

1. a better plan for the Public / Private aspects of the API
2. a shim for ensuring backwards compatibility for existing users

Although, I don't want this PR to get too long...

**Comment by HaoZeke** [comment]
@pearu and @melissawm, this should be good to go now at last.

**Comment by melissawm** [comment]
There's a few tests failures now

**Comment by HaoZeke** [comment]
This should now pass all the tests. As it stands it might be a bit large, but it can be broken conceptually into two separate parts (was implemented separately too):

1. The deprecation helper --> Part of this is kanged, and simplified for the most part; the main reason this is needed is that projects often call `from numpy.f2py import crackfortran`
2. The actual refactoring --> This involved the renaming of certain symbols to ensure tests pass

As for adding to the `undocumented` dictionary, not happy about it, but in an ideal world that entire file can be gone in 3 releases...

**Comment by HaoZeke** [comment]
Things to test:
- [ ] `f90wrap`
- [ ] `scipy`

**Comment by HaoZeke** [comment]
# `module` deprecation utility

This is a paraphrased version of the [Cirq implementation document](https://docs.google.com/document/d/1tQTDla0RNuxB2rMfeQ5YqR6uO-qsqpwz78IlQFyCNlQ/edit). The implementation relies heavily on Python's [loading mechanisms](https://docs.python.org/3/reference/import.html#loading) using `importlib`.

At a high level:
- Intercept call to old module (by adding `DeprecatedModuleFinder` to `sys.meta_path`)
  - Check that the new module is valid
- Call `DeprecatedModuleLoader` which in turn wraps `exec_method` to ensure both paths resolve to the same `ModuleSpec` in `sys.modules`
  - Aliases are added to `sys.modules` cache recursively

The `ModuleSpec` specification ([PEP 451](https://www.python.org/dev/peps/pep-0451/)) ensures that the the module is not actually loaded until needed.

The interface to this is simply:

```python
deprecated_submodule(new_module_name = "numpy.f2py.stds.auxfuncs",
                     old_parent = "numpy.f2py",
                     old_child = "auxfuncs",
                     deadline = "1.25")
```

- It will raise after `1.25`
- No spurious files to purge

Additionally, this cleanly handles the case of where modules may be doubly executed otherwise:

```python
import numpy as np
from np import linalg # error
sys.modules['np'] = np
from np import linalg # works
from numpy import linalg # executes linalg again
```

This situation can be (and has been) handled in by using the `Loader` mechanism.

## Alternatives
### Stub file generation
This is the approach taken by `scikit-learn` as [demonstrated here](https://github.com/thomasjpfan/scikit-learn/blob/235608380799995fa5ef3bff37524212b0480af5/sklearn/_build_utils/deprecated_modules.py) (since removed), as noted by @rgommers.

- Stub files are generated during the installation
- These `import` the entire public API
- It does not seem to extend to being backwards compatible with only loading some part of the old API (e.g. `from x import y`) as the entire module is executed
- Unclear in a first read if this supports `as` (e.g. `import y as z`)

## Biased comparison
IMO, it is cleaner to not generate files which are not needed and can easily be done away with. Additionally, it feels more pythonic to use the standard library's import mechanisms.

Finally, the question of aliases and other double execution of modules are also not cleanly resolved otherwise.

File: numpy/f2py/utils/deprecation_helpers.py [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+    assert re.match(DEADLINE_REGEX, deadline),\
+           "deadline should match X.Y"
+    assert (float(f"{dep_major_version}.{dep_minor_version}") >
+            float(f"{np_major_version}.{np_minor_version}")),\
+            f"""For replacing {old_module_name} with {new_module_name}..
+            Deadline: {dep_major_version}.{dep_minor_version}
+            Got: {np_major_version}.{np_minor_version}"""
+
+
+def deprecated_submodule(*, new_module_name, old_parent, old_child, deadline):
```
[the remark this answers is not in the document]
Comment (HaoZeke) [reply numpy/f2py/utils/deprecation_helpers.py:48]: I have added a comment describing the high level structure of this mechanism for further discussion.

**Comment by rgommers** [comment]
> IMO, it is cleaner to not generate files which are not needed and can easily be done away with. Additionally, it feels more pythonic to use the standard library's import mechanisms.

This sounds nice in theory, but `importlib` has been extremely buggy, and this code is too complex. I have a very strong preference for not merging it. 

The scikit-learn and SciPy (which does a similar module-level `__getattr__` thing, just not with codegen) is much easier to understand.

> 1. The deprecation helper --> Part of this is kanged, and simplified for the most part; the main reason this is needed is that projects often call `from numpy.f2py import crackfortran`

This makes things easier: how about you just commit a single `crackfortran.py` file with a `__getattr__`, and be done with it? Nothing else here seems of much use for third-party code and it's not public API, so you're fine just refactoring it without providing shims.

**Comment by HaoZeke** [comment]
Closing as discussed in https://github.com/numpy/numpy/issues/20480

**Comment by HaoZeke** [comment]
Closing as discussed in https://github.com/numpy/numpy/issues/20480

**Pull request closed without merging**